src/tree/updater_gpu_hist.cu

```diff
@@ -1444,23 +1453,64 @@ class GPUHistMakerSpecialised{
                                           column_sampling_seed));
         });
 
-    // Find the cuts.
     monitor_.StartCuda("Quantiles");
-    common::DeviceSketch(batch, *info_, param_, &hmat_, hist_maker_param_.gpu_batch_nrows);
-    n_bins_ = hmat_.row_ptr.back();
+    // Create the quantile sketches for the dmatrix
+    std::vector<common::HistCutMatrix::WXQSketch> sketches;
+    size_t row_stride = common::DeviceSketch(param_, hist_maker_param_.gpu_batch_nrows,
+                                             dmat, &sketches);
     monitor_.StopCuda("Quantiles");
+
+    // Initialize hmat_ for the entire data set
+    hmat_.Init(&sketches, param_.max_bin);
+
+    n_bins_ = hmat_.row_ptr.back();
+
     auto is_dense = info_->num_nonzero_ == info_->num_row_ * info_->num_col_;
 
-    monitor_.StartCuda("BinningCompression");
+    // Init global data for each shard
+    monitor_.StartCuda("InitCompressedData");
     dh::ExecuteIndexShards(
         &shards_,
         [&](int idx, std::unique_ptr<DeviceShard<GradientSumT>>& shard) {
           dh::safe_cuda(cudaSetDevice(shard->device_id));
-          shard->InitCompressedData(hmat_, batch, is_dense);
+          shard->InitCompressedData(hmat_, row_stride, is_dense);
         });
+    monitor_.StopCuda("InitCompressedData");
+
+    monitor_.StartCuda("BinningCompression");
+    size_t processed_size = 0;
+    // TODO(sriramch): Think of a way to utilize *all* the GPUs to build the
+    // compressed bins
+    for (const auto &batch : dmat->GetRowBatches()) {
+      std::vector<size_t> shard_allocations(shards_.size());
+      // Distribute the rows in this batch to the different shards
+      for (size_t i = 0; i < shards_.size(); ++i) {
+        // Does this batch pertain to me?
+        int64_t num_elems = processed_size + batch.Size() - shards_[i]->row_begin_idx;
```

Comment:
As before be careful using the word 'elems'.

Reply:
done